src/client_side_request.cc

```diff
@@ -1637,6 +1631,55 @@ ClientHttpRequest::loggingEntry(StoreEntry *newEntry)
         loggingEntry_->lock("ClientHttpRequest::loggingEntry");
 }
 
+void
+ClientHttpRequest::initRequest(HttpRequest *aRequest)
+{
+    assert(aRequest);
+    assert(!request);
+    // TODO: also reset uri if possible
+    resetRequest(aRequest, false);
```

Comment:
This is the only place where we are _not_ resetting the URI. This is also the only place where we are not actually _resetting_ the request (we are setting the first request, not replacing the old one). This resetRequest() abuse has bothered me before but this new boolean parameter is the last straw -- we should split resetRequest() into two methods:

1. resetRequest(newRequest) -- a public method called to replace the old request with a new one. Always resets the `uri` member as well. Very similar to the current `resetRequest(..., true)` usage.
1. assignRequest(aRequest) -- a private method that performs the common tasks of initRequest() and resetRequest(). It assigns the request and calls setLogUriToRequestUri() while asserting that the current request is nil (either because it did not exist or because it was just cleared by the caller).